pkg/services/ngalert/api/tooling/definitions/cortex-ruler.go

```diff
@@ -396,3 +394,70 @@ type GettableGrafanaRule struct {
 	Provenance      Provenance          `json:"provenance,omitempty" yaml:"provenance,omitempty"`
 	IsPaused        bool                `json:"is_paused" yaml:"is_paused"`
 }
+
+// AlertQuery represents a single query associated with an alert definition.
+type AlertQuery struct {
+	// RefID is the unique identifier of the query, set by the frontend call.
+	RefID string `json:"refId"`
+	// QueryType is an optional identifier for the type of query.
+	// It can be used to distinguish different types of queries.
+	QueryType string `json:"queryType"`
+	// RelativeTimeRange is the relative Start and End of the query as sent by the frontend.
+	RelativeTimeRange RelativeTimeRange `json:"relativeTimeRange"`
+
+	// Grafana data source unique identifier; it should be '__expr__' for a Server Side Expression operation.
+	DatasourceUID string `json:"datasourceUid"`
+
+	// JSON is the raw JSON query and includes the above properties as well as custom properties.
+	Model json.RawMessage `json:"model"`
+}
+
+// RelativeTimeRange is the per query start and end time
+// for requests.
+type RelativeTimeRange struct {
+	From Duration `json:"from" yaml:"from"`
+	To   Duration `json:"to" yaml:"to"`
+}
+
+// Duration is a type used for marshalling durations.
+type Duration time.Duration
```

Comment:
Remainder is copied verbatim from models :+1: 

I wonder - does it _ever_ make sense to have any shared types at all between definitions and other packages? Duration is pretty abstract, for example. Perhaps `definitions` could have one dependency package for really basic things like this? `model` could depend on it too, and you still have no dependency on either direction between `definitions` and `model`, so there isn't a breach of API/domain separation. Just a thought.

Reply:
ideally, I would prefer only to share structs from the base library between layers. We can make some base models that are recognized as a fundamental models but in this case we will have to be very careful to not leak any layer-specific logic to that package, which I think is hard.

Duration is defined here because of the custom marshaling, which is not required in, say database layer (the fact that we use custom marshaling in the storage layer I think is not a good thing). By having separate types for the duration in API layer and storage or domain layer we can avoid leaking logic specific from one layer to another layer, like marshaling in this case.